README.md
# RegularizationTools

*A Julia package to perform Tikhonov regularization for small to moderate size problems.*

| **Documentation**                                           |  **Citations** |
|:-------------------------------------------------------------------------------:|:-------------------------------------------------------------------------------------------------------------------------------------------:|
| [![][docs-stable-img]][docs-stable-url] | [![DOI](https://img.shields.io/badge/DOI-10.5194%2Famt--14--7909--2021-blue)](https://doi.org/10.5194/amt-14-7909-2021)  |

# Installation

The package can be installed from the Julia package prompt with

```julia
julia> ]add RegularizationTools
```

The closing square bracket switches to the package manager interface and the ```add``` command installs the package and any missing dependencies. To return to the Julia REPL hit the ```delete``` key.

To load the package run

```julia
julia> using RegularizationTools
```

## Documentation

- [**STABLE**][docs-stable-url] &mdash; **documentation of the most recently tagged version.**

## Project Status
The package is being developed for, Julia `1.5` and above.

## Citations

*If you use this package and publish your work, please cite:*

Petters, M. D.: Revisiting matrix-based inversion of scanning mobility particle sizer (SMPS) and humidified tandem differential mobility analyzer (HTDMA) data, Atmos. Meas. Tech., 14, 7909–7928, https://doi.org/10.5194/amt-14-7909-2021, 2021.

This work was supported by the United States Department of Energy, Office of Science, Biological and Environment Research, Grant number DE-SC0021074.

[docs-dev-img]: https://img.shields.io/badge/docs-dev-blue.svg
[docs-dev-url]: https://mdpetters.github.io/RegularizationTools.jl/latest/

[docs-stable-img]: https://img.shields.io/badge/docs-stable-blue.svg
[docs-stable-url]: https://mdpetters.github.io/RegularizationTools.jl/stable/


NEWS.md
# RegularizationTools.jl NEWS

#### Notes on release changes and ongoing development
---
## v0.5.2
- change derivative matrix function call to Γ(m::Int, order::Int)
- remove Matrix Depot as dependency

## v0.5.0
- Significant performance gain in "setupRegularizationProblem" by switching from GSVD to QR factorization approach
- Update documentation
- Update tests to be independent of random number generation
- Update Project and Manifest to be current. Added CSV and DataFrames as deps for unit tests
- Explicitly account for the unregularized component of x which is not affected by the regularization scheme (fix for issue #7 raised by Jonathan Stickel). Note that the bug/issue affected solutions for very larger regularization parameters, but had no/neglible effect on calculations near the optimal solution for regularization. 
- Update code for smoothing matrix generation (base on example in issue #7 contributed by Jonathan Stickel). The updated code also directly allows computation of arbitrary order of the smoothing matrix.
- Added smoothing example from issue #7 to examples folder

## v0.4.1
- Minor release associated with julia 1.6.0 
- Update Project and Manifest
- Make tests of invert function resilient against changes in random number generator
- Change order in scattering example to 2

## v0.4.0
- Add high-level API invert function to simplify notation

## v0.3.1
- Add memoization
- Merge CompatHelper: add new compat entry for "Memoize" at version "0.4"
- Fix various bugs related to non-square matrices

## v0.3.0
- Add abstract interface to generate design matrix from a forward model.
- Add documentation and examples how to use the interface
- Add fallback in solve function in case cholesky factorization fails.
- Fix bug when creating L matrix for non-square design matrices.

## v0.2.0
- Add constraint minimization solver to enfore upper and lower bound for some problems.
- Add documentation.

## v0.1.1
- Fix standard form conversion error. Solution is now based on generalized SVD. 
- Touch up documentation.
- RegularizationTools is now part of the general registry

## v0.1.0
- Initial Release


test/invert.jl
include("mie_functions.jl")
include("size_functions.jl")

function get_domainfunction(ref)   
    function f(node::Domain)
        s, x, q = node.s, node.x, node.q
        Dp, λ = (@_ map(_[2], s)), q[2]
       
        Q = @match q[1][1:3] begin 
            "βs" => @_ map(Qsca(π*Dp[_]*1e-9/λ, ref), 1:length(Dp))
            "βa" => @_ map(Qabs(π*Dp[_]*1e-9/λ, ref), 1:length(Dp))
            _ => trow("error")
        end
        mapreduce(+, 1:length(Dp)) do i
            π/4.0 *(x[i]*1e6)*Q[i]*(Dp[i]*1e-6)^2.0*1e6
        end
    end
end

Dp, N = lognormal([[100.0, 500.0, 1.4]], d1 = 100.0, d2 = 1000.0, bins = 8)
λs = [300e-9, 500e-9, 900e-9, 1200e-9, 1500e-9]

s = zip(["Dp$i" for i = 1:length(Dp)],Dp) |> collect
q1 = zip(["βs$i" for i in Int.(round.(λs*1e9, digits=0))], λs) |> collect
q2 = zip(["βa$i" for i in Int.(round.(λs*1e9, digits=0))], λs) |> collect
q = [q1; q2] 

ref = complex(1.8, 0.01)
f = get_domainfunction(ref)
A = designmatrix(s, q, f) 

e = [-0.4, 0.12, 1.26, -1.42, 0.3, 0.8, 0.15, 0.04, -0.68, -0.61]
b = A*N .+ 0.001*(A*N) .* e

lb, ub, ε, x₀ = zeros(8), zeros(8) .+ 50.0, 0.02, 0.5*N

xλ1 = @> setupRegularizationProblem(A,2) solve(b, alg=:gcv_svd) getfield(:x)
xλ2 = invert(A, b, Lₖ(2); alg = :gcv_svd)
@test xλ1 ≈ xλ2

inrange(x, y, δ) = ifelse((x < y + δ) & (x > y - δ), true, false)

xλ = invert(A, b, Lₖ(2); alg = :gcv_svd, λ₁ = 0.1)
@test inrange(sum(xλ), 106.0, 2.0) == true
xλ = invert(A, b, LₖB(2,lb,ub); alg = :L_curve, λ₁ = 0.1)
@test inrange(sum(xλ), 145.0, 2.0) == true
xλ = invert(A, b, Lₖx₀(2, x₀); alg = :L_curve, λ₁ = 0.1)
@test inrange(sum(xλ), 76.0, 2.0) == true
xλ = invert(A, b, Lₖx₀(2, x₀); alg = :gcv_tr)
@test inrange(sum(xλ), 109.0, 2.0) == true
xλ = invert(A, b, Lₖx₀(2, x₀); alg = :gcv_svd)
@test inrange(sum(xλ), 109.0, 2.0) == true
xλ = invert(A, b, Lₖx₀B(2,x₀,lb,ub); alg = :gcv_svd)
@test inrange(sum(xλ), 109.0, 2.0) == true
xλ = invert(A, b, LₖDₓ(2, ε); alg = :gcv_svd)
@test inrange(sum(xλ), 109.0, 2.0) == true
xλ = invert(A, b, LₖDₓB(2, ε, lb, ub); alg = :gcv_svd)
@test inrange(sum(xλ), 109.0, 2.0) == true
xλ = invert(A, b, Lₖx₀Dₓ(2,x₀, ε); alg = :gcv_svd)
@test inrange(sum(xλ), 109.0, 2.0) == true
xλ = invert(A, b, Lₖx₀DₓB(2, x₀, ε, lb, ub); alg = :gcv_svd)
@test inrange(sum(xλ), 109.0, 2.0) == true


test/solvers.jl
r = mdopen("shaw", 100, false)
A, x = r.A, r.x
rn = CSV.read("random_numbers.csv", DataFrame)
y = A  * x
b = y + 0.1y .* rn.x
x₀ = 0.5x

Ψ = setupRegularizationProblem(A,1)
λopt = @> solve(Ψ, b, alg=:gcv_svd) getfield(:λ)
@test round(λopt, digits = 3) ≈ 2.396
λopt = @> solve(Ψ, b, alg=:gcv_tr) getfield(:λ)
@test round(λopt, digits = 3) ≈ 2.396
λopt = @> solve(Ψ, b, alg =:L_curve) getfield(:λ)
@test round(λopt, digits = 3) ≈ 1.681
λopt = @> solve(Ψ, b, x₀, alg =:L_curve) getfield(:λ)
@test round(λopt, digits = 3) ≈ 1.725


test/runtests.jl
using RegularizationTools
using MatrixDepot
using Memoize
using MLStyle
using Random
using Test
using Underscores
using LinearAlgebra
using Lazy
using CSV
using DataFrames

@testset "RegularizationTools.jl" begin
    include("standard_form.jl")  # Test standard form transformation
    include("solvers.jl")        # Test optimization
    include("validators.jl")     # Test validators
    include("domainfunctions.jl") 
    include("invert.jl")
end


src/invert.jl
@doc raw"""
    function invert(A::Matrix, b::Vector, method::InverseMethod; kwargs...)

High-level API function to perform Tikhonov inversion. The function used 
the algebraic data type InverseMethod

```julia
    @data InverseMethod begin 
        Lₖ(Int)                            # Pure Tikhonov
        Lₖx₀(Int,Vector)                   # with initial guess
        LₖB(Int,Vector,Vector)             # with bounds
        Lₖx₀B(Int,Vector,Vector,Vector)    # with initial guess + bounds
        LₖDₓ(Int,Float64)                  # with filter 
        Lₖx₀Dₓ(Int,Vector,Float64)         # with initial guess + filter 
        LₖDₓB(Int,Float64,Vector,Vector)   # with filter + bound
        Lₖx₀DₓB(Int,Vector,Float64,Vector,Vector) # with initial guess + filter + bound
    end
```

to define the problem and then dispatches to the correct method. The kwargs... are passed to the solve function
For example the standard way to perform second order regularization is 

```julia
xλ = @> setupRegularizationProblem(A, 2) solve(b) getfield(:x)
```

this can alternativel written as 

```julia
invert(A, b, Lₖ(2))
```

where ```Lₖ(2)``` denotes the second order method. The method nomenclature is
```Lₖ``` for regularization order, ```B``` for bounded search, ```x₀``` for 
initial condition, and ```Dₓ``` for the Huckle and Sedlacek (2012) two-step data based 
regularization. The method data type takes hyper parameters to initialize the search. 
Examples of method initializations are

```julia
# Hyper parameters 
k: order, lb: low bound, ub: upper bound, ε: noise level, x₀: initial guess
k, lb, ub, ε, x₀ = 2, zeros(8), zeros(8) .+ 50.0, 0.02, 0.5*N

xλ = invert(A, b, Lₖ(k); alg = :gcv_tr, λ₁ = 0.1)
xλ = invert(A, b, Lₖ(k); alg = :gcv_svd, λ₁ = 0.1)
xλ = invert(A, b, LₖB(k, lb, ub); alg = :L_curve, λ₁ = 0.1)
xλ = invert(A, b, Lₖx₀(k, x₀); alg = :L_curve, λ₁ = 0.1)
xλ = invert(A, b, Lₖx₀(k, x₀); alg = :gcv_tr)
xλ = invert(A, b, Lₖx₀(k, x₀); alg = :gcv_svd)
xλ = invert(A, b, Lₖx₀B(k, x₀, lb, ub); alg = :gcv_svd)
xλ = invert(A, b, LₖDₓ(k, ε); alg = :gcv_svd)
xλ = invert(A, b, LₖDₓB(k, ε, lb, ub); alg = :gcv_svd)
xλ = invert(A, b, Lₖx₀Dₓ(k, x₀, ε); alg = :gcv_svd)
xλ = invert(A, b, Lₖx₀DₓB(k, x₀, ε, lb, ub); alg = :gcv_svd)
```
"""    
function invert(A::Matrix, b::Vector, method::InverseMethod; kwargs...)
    n,m = size(A)
    
    res = @match method begin
        Lₖ(k) => @as p begin
            setupRegularizationProblem(A, k)
            solve(p, b; kwargs...)    
            getfield(p, :x)
        end
        Lₖx₀(k, x₀) => @as p begin
            setupRegularizationProblem(A, k)
            solve(p, b, x₀; kwargs...)      
            getfield(p, :x)
        end
        LₖB(k, lb, ub)  => @as p begin
            setupRegularizationProblem(A, k)
            solve(p, b, lb, ub; kwargs...)     
            getfield(p, :x)
        end
        Lₖx₀B(k, x₀, lb, ub) => @as p begin
            setupRegularizationProblem(A, k)
            solve(p, b, x₀, lb, ub; kwargs...) 
            getfield(p, :x)
        end
        LₖDₓ(k, ε) => begin
            x̂ = @as p begin
                setupRegularizationProblem(A, k)
                solve(p, b; kwargs...) 
                getfield(p, :x)
            end 
            x̂[x̂ .< ε] .= ε 
            ψ = @>> Γ(m, k)*Diagonal(x̂)^(-1) setupRegularizationProblem(A)
            @as p begin
                solve(ψ, b; kwargs...) 
                getfield(p, :x)
            end
        end
        LₖDₓB(k, ε, lb, ub) => begin
            x̂ = @as p begin
                setupRegularizationProblem(A, k)
                solve(p, b; kwargs...) 
                getfield(p, :x)
            end 
            x̂[x̂ .< ε] .= ε
            ψ = @>> Γ(m, k)*Diagonal(x̂)^(-1) setupRegularizationProblem(A)
            @as p begin
                solve(ψ, b, x̂, lb, ub; kwargs...) 
                getfield(p, :x)
            end 
        end
        Lₖx₀Dₓ(k, x₀, ε) => begin
            x̂ = @as p begin
                setupRegularizationProblem(A, k)
                solve(p, b, x₀; kwargs...) 
                getfield(p, :x)
            end 
            x̂[x̂ .< ε] .= ε
            ψ = @>> Γ(m, k)*Diagonal(x̂)^(-1) setupRegularizationProblem(A)
            @as p begin
                solve(ψ, b; kwargs...) 
                getfield(p, :x)
            end 
        end
        Lₖx₀DₓB(k, x₀, ε, lb, ub) => begin
            x̂ = @as p begin
                setupRegularizationProblem(A, k)
                solve(p, b, x₀; kwargs...) 
                getfield(p, :x)
            end 
            x̂[x̂ .< ε] .= ε
            ψ = @>> Γ(m, k)*Diagonal(x̂)^(-1) setupRegularizationProblem(A)
            @as p begin
                solve(ψ, b, x̂, lb, ub; kwargs...) 
                getfield(p, :x)
            end 
        end
        _ => throw("Unknown Method")
    end
end


src/solvers.jl
@doc raw"""
    Γ(m::Int, order::Int)

Return the smoothing matrix L for Tikhonov regularization of a system of size `m`.  Order
can be 0, 1, ..., n. Code to generate matrix is based on the suggestion posted in Issue #7,
which was inspired by Eilers, P. H. C. (2003). Analytical Chemistry, 75(14),
3631–3636. (Supporting Information).

```julia
L = Γ(m, 1)
```
"""
@memoize function Γ(m::Int, order::Int)
    if order == 0
        return Array{Float64}(LinearAlgebra.I, (m, m))
    end

    return diff(Γ(m, order-1), dims=1)
end

function zot(A::AbstractMatrix, λ::AbstractFloat)
    a = deepcopy(A)
    n = size(A'A, 1)
    for i = 1:n
        @inbounds a[i, i] += λ
    end
    return a
end

@doc raw"""
    to_standard_form(Ψ::RegularizationProblem, b::AbstractVector)

Converts vector b to standard form using (Hansen, 1998)

Example Usage (Regular Syntax)
```julia
b̄ = to_standard_form(Ψ, b)
```

Example Usage (Lazy Syntax)
```julia
b̄ = @>> b to_standard_form(Ψ)
```
"""
to_standard_form(Ψ::RegularizationProblem, b::AbstractVector) = b - Ψ.A*Ψ.K₀T⁻¹H₀ᵀ*b

@doc raw"""
    to_standard_form(Ψ::RegularizationProblem, b::AbstractVector, x₀::AbstractVector)

Converts vector b and x₀ to standard form using (Hansen, 1998)

Example Usage (Regular Syntax)
```julia
b̄ = to_standard_form(Ψ, b, x₀)
```
"""
to_standard_form(Ψ::RegularizationProblem, b::AbstractVector, x₀::AbstractVector) =
    b - Ψ.A*Ψ.K₀T⁻¹H₀ᵀ*b, Ψ.L * x₀

@doc raw"""
    to_general_form(Ψ::RegularizationProblem, b::AbstractVector, x̄::AbstractVector)

Converts solution ``\bar {\rm x}`` computed in standard form back to general form 
``{\rm x}`` using (Hansen, 1998).

```math
{\rm x}={\rm {\bf L^{+}_A}\bar{x} + x\_0}
```

where the matrices and vectors are defined in [RegularizationProblem](@ref)

Example Usage (Regular Syntax)
```julia
x = to_general_form(Ψ, b, x̄) 
```

Example Usage (Lazy Syntax)
```julia
x = @>> x̄ to_general_form(Ψ, b) 
```
"""
function to_general_form(Ψ::RegularizationProblem, b::AbstractVector, x̄::AbstractVector) 
	x = Ψ.L⁺ * x̄  + Ψ.K₀T⁻¹H₀ᵀ*(b - Ψ.A*Ψ.L⁺*x̄ ) 
end

@doc raw"""
    solve(Ψ::RegularizationProblem, b̄::AbstractVector, λ::AbstractFloat)

Compute the Tikhonov solution for problem Ψ in standard form for regularization parameter λ
and using zero as initial guess. Returns a vector ``\rm {\bar x}_\lambda``. 

```math
{\rm x_{\lambda}}=\left({\rm {\bf \bar A}^{T}}{\rm {\bf \bar A}}+\lambda^{2}{\rm {\bf I}}\right)^{-1} 
{\rm {\bf {\bar A}}^{T}}{\rm {\bar b}} 
```
Example Usage (Standard Syntax)
```julia
# A is a Matrix and b is a response vector. 
Ψ = setupRegularizationProblem(A, 1)     # Setup problem
b̄ = to_standard_form(Ψ, b)               # Convert to standard form
x̄ = solve(A, b̄, 0.5)                     # Solve the equation
x = to_general_form(Ψ, b, x̄)             # Convert back to general form
```

Example Usage (Lazy Syntax)
```julia
# A is a Matrix and b is a response vector. 
Ψ = setupRegularizationProblem(A, 1)     # Setup problem
b̄ = @>> b to_standard_form(Ψ)            # Convert to standard form
x̄ = solve(A, b̄, 0.5)                     # Solve the equation
x = @>> x̄ to_general_form(Ψ, b)          # Convert back to general form
```
"""
solve(Ψ::RegularizationProblem, b̄::AbstractVector, λ::AbstractFloat) = try
    cholesky!(Hermitian(zot(Ψ.ĀĀ, λ^2.0))) \ (Ψ.Ā' * b̄)
catch
     zot(Ψ.ĀĀ, λ^2.0) \ (Ψ.Ā' * b̄)
end

@doc raw"""
    solve(Ψ::RegularizationProblem, b̄::AbstractVector, x̄₀::AbstractVector, λ::AbstractFloat)

Compute the Tikhonov solution for problem Ψ in standard form for regularization parameter λ
and using x̄₀ as initial guess. 

```math
{\rm x_{\lambda}}=\left({\rm {\bf \bar A}^{T}}{\rm {\bf \bar A}}+\lambda^{2}{\rm {\bf I}}\right)^{-1} 
\left({\rm {\bf {\bar A}}^{T}}{\rm {\bar b}} + \lambda^2 {\rm {\bar x}}_0 \right)
```

Example Usage (Standard Syntax)
```julia
# A is a Matrix and b is a response vector. 
Ψ = setupRegularizationProblem(A, 2)     # Setup problem
b̄, x̄₀ = to_standard_form(Ψ, b, x₀)       # Convert to standard form
x̄ = solve(A, b̄, x̄₀, 0.5)                 # Solve the equation
x = to_general_form(Ψ, b, x̄)             # Convert back to general form
```
"""
solve(Ψ::RegularizationProblem, b̄::AbstractVector, x̄₀::AbstractVector, λ::AbstractFloat) = try
    cholesky!(Hermitian(zot(Ψ.ĀĀ, λ^2.0))) \ (Ψ.Ā' * b̄ + λ^2.0 * x̄₀)
catch
    zot(Ψ.ĀĀ, λ^2.0) \ (Ψ.Ā' * b̄ + λ^2.0 * x̄₀)
end

@doc raw"""
    function solve(
        Ψ::RegularizationProblem,
        b::AbstractVector;
        alg = :gcv_svd,
        λ₁ = 0.0001,
        λ₂ = 1000.0,
    )

Find the optimum regularization parameter λ between [λ₁, λ₂] using the algorithm alg. Choices
for algorithms are
```    
    :gcv_tr - generalized cross validation using the trace formulation (slow)
    :gcv_svd - generalized cross validation using the SVD decomposition (fast)
    :L_curve - L-curve algorithm 
```

!!! tip
    The gcv\_svd algorithm is fastest and most stable. The L\_curve algorithn is sensitive to the upper 
    and lower bound. Specify narrow upper and lower bounds to obtain a good solution.

The solve function takes the original data, converts it to standard form, performs the search
within the specified bounds and returns a [RegularizatedSolution](@ref)

Example Usage (Standard Syntax)
```julia
# A is a Matrix and b is a response vector. 
Ψ = setupRegularizationProblem(A, 2)     # Setup problem
sol = solve(Ψ, b)                        # Solve it
```

Example Usage (Lazy Syntax)
```julia
# A is a Matrix and b is a response vector. 
sol = @> setupRegularizationProblem(A, 1) solve(b)
```
"""
function solve(
    Ψ::RegularizationProblem,
    b::AbstractVector;
    alg = :gcv_svd,
    λ₁ = 0.0001,
    λ₂ = 1000.0,
)
    b̄ = @>> b to_standard_form(Ψ)
    L1, L2, κ = Lcurve_functions(Ψ, b̄)

    solution = @match alg begin
        :gcv_tr  => @_ optimize(gcv_tr(Ψ, b̄, _), λ₁, λ₂) 
        :gcv_svd => @_ optimize(gcv_svd(Ψ, b̄, _), λ₁, λ₂) 
        :L_curve => @_ optimize(1.0 - κ(_), λ₁, λ₂) 
        _ => throw("Unknown algorithm, use :gcv_tr, :gcv_svd, or :L_curve")
    end

    λ = @> solution Optim.minimizer
    x̄ = solve(Ψ, b̄, λ)
    x = @>> x̄ to_general_form(Ψ, b) 
     
    return RegularizedSolution(x, λ, solution)
end

@doc raw"""
    function solve(
        Ψ::RegularizationProblem,
        b::AbstractVector,
        x₀::AbstractVector;
        alg = :gcv_svd,
        λ₁ = 0.0001,
        λ₂ = 1000.0,
    )

Same as above, but includes an initial guess x₀. Example Usage (Lazy Syntax)
```julia
# A is a Matrix and b is a response vector. 
sol = @> setupRegularizationProblem(A, 1) solve(b, x₀, alg = :L_curve, λ₂ = 10.0)
```
"""
function solve(
    Ψ::RegularizationProblem,
    b::AbstractVector,
    x₀::AbstractVector;
    alg = :gcv_svd,
    λ₁ = 0.0001,
    λ₂ = 1000.0,
)
    b̄ = @>> b to_standard_form(Ψ)
    x̄₀ = Ψ.L * x₀
    L1, L2, κ = Lcurve_functions(Ψ, b̄, x̄₀)

    solution = @match alg begin
        :gcv_tr  => @_ optimize(gcv_tr(Ψ, b̄, x̄₀, _), λ₁, λ₂) 
        :gcv_svd => @_ optimize(gcv_svd(Ψ, b̄, x̄₀, _), λ₁, λ₂)
        :L_curve => @_ optimize(1.0 - κ(_), λ₁, λ₂) 
        _ => throw("Unknown algorithm, use :gcv_tr, :gcv_svd, or :L_curve")
    end

    λ = @> solution Optim.minimizer
    x̄ = solve(Ψ, b̄, x̄₀, λ)
    x = @>> x̄ to_general_form(Ψ, b)   

    return RegularizedSolution(x, λ, solution)
end

@doc raw"""
    function solve(
        Ψ::RegularizationProblem, 
        b::AbstractVector,
        lower::AbstractVector, 
        upper::AbstractVector;
        kwargs...
    )

Constraint minimization of [RegularizationProblem](@ref) Ψ, with observations b and
upper and lower bounds for each xᵢ.

The function computes the algebraic solution using ```solve(Ψ, b; kwargs...)```, truncates the
solution at the upper and lower bounds and uses this solution as initial condition for
the minimization problem using a Least Squares numerical solver. The returned solution
is using the regularization parameter λ obtained from the algebraic solution.
"""
function solve(
    Ψ::RegularizationProblem, 
    b::AbstractVector,
    lower::AbstractVector, 
    upper::AbstractVector;
    kwargs...
)
    return solve_numeric(Ψ, b, solve(Ψ, b; kwargs...), lower, upper)
end

@doc raw"""
    function solve(
        Ψ::RegularizationProblem, 
        b::AbstractVector,
        x₀::AbstractVector,
        lower::AbstractVector, 
        upper::AbstractVector;
        kwargs...
    )


Constraint minimization of [RegularizationProblem](@ref) Ψ, with observations b, intial 
guess x₀ and upper and lower bounds for each xᵢ.

The function computes the algebraic solution using ```solve(Ψ, b; kwargs...)```, truncates the
solution at the upper and lower bounds and uses this solution as initial condition for
the minimization problem using a Least Squares numerical solver. The returned solution
is using the regularization parameter λ obtained from the algebraic solution.
"""    
function solve(
    Ψ::RegularizationProblem, 
    b::AbstractVector,
    x₀::AbstractVector,
    lower::AbstractVector, 
    upper::AbstractVector;
    kwargs...
)
    return solve_numeric(Ψ, b, solve(Ψ, b, x₀; kwargs...), lower, upper)
end


function solve_numeric(
    Ψ::RegularizationProblem, 
    b::AbstractVector, 
    xλ::RegularizedSolution,
    lower::AbstractVector,
    upper::AbstractVector
)
    λ = xλ.λ  
    xᵢ = xλ.x
    xᵢ[xᵢ .< lower] .= lower[xᵢ .< lower]
    xᵢ[xᵢ .> upper] .= upper[xᵢ .> upper]
    
    LᵀL = Ψ.L'*Ψ.L
    n = size(LᵀL,1)
    
    function f!(out, x)
        out[1] = norm(Ψ.A*x - b)^2.0 + λ^2.0*norm(LᵀL*x)^2.0
    end
    
    function g!(out, x)
        ot = Ψ.A'*(Ψ.A*x - b) + λ^2.0*LᵀL*x
        [out[i] = 2.0*ot[i] for i = 1:n]
    end
    
    LLSQ = LeastSquaresProblem(
        x = xᵢ, 
        f! = f!, 
        g! = g!,
        output_length=n
    )
   
    r = optimize!(
        LLSQ, 
        Dogleg(LeastSquaresOptim.QR()), 
        lower = lower, 
        upper = upper,
        x_tol=1e-10
    ) 
    return return RegularizedSolution(r.minimizer, λ, r)
end


@doc raw"""
    setupRegularizationProblem(A::AbstractMatrix, order::Int)

Precompute matrices to initialize Reguluarization Problem based on design matrix A and 
zeroth, first, or second order difference operator. See Hanson (1998) and source code
for details.

Example Usage
```julia
Ψ = setupRegularizationProblem(A, 0) # zeroth order problem
Ψ = setupRegularizationProblem(A, 2) # second order problem
```
"""
setupRegularizationProblem(A::AbstractMatrix, order::Int) = 
    setupRegularizationProblem(A, Γ(size(A,2), order))

@doc raw"""
    setupRegularizationProblem(A::AbstractMatrix, L::AbstractMatrix)

Precompute matrices to initialize Reguluarization Problem based on design matrix and 
Tikhonov smoothing matrix. See Hansen (1998, Eq. 2.35)

Example Usage
```julia
Ψ = setupRegularizationProblem(A, L) 
```
"""
@memoize function setupRegularizationProblem(A::AbstractMatrix, L::AbstractMatrix)
    p, n = size(L)
    Iₙ = Matrix{Float64}(I, n, n) 
    Iₚ = Matrix{Float64}(I, p, p)
	Q,R = qr(L')
	K0 = Q[:,p+1:end]
	H,T = qr(A*K0)
	H0 = H[:,1:n-p]
	K₀T⁻¹H₀ᵀ = K0*T^-1*H0'
	L⁺ = pinv(L)
	L⁺ₐ = (Iₙ - K₀T⁻¹H₀ᵀ*A)*L⁺
    Ā = A*L⁺ₐ

    RegularizationProblem(
        Ā,
        A,
        L,
        Ā'Ā,
        Ā',
        svd(Ā),
        Iₙ,
        Iₚ,
        L⁺ₐ,
		L⁺,
		K₀T⁻¹H₀ᵀ	
    )
end


examples/smoothing.jl
# Example from Jonathan Stickel, with code posted in issue #7
# The raised issue has been addressed in version 0.5.0 
import RegularizationTools
const RT = RegularizationTools
import LinearAlgebra
const LA = LinearAlgebra
using PyPlot

# create simulated data
npts = 100
xmin = 0.0
xspan = 3/2*π
x = range(xmin, xmin+xspan, length=npts)
yt = sin.(x)
stdev = 1e-1*maximum(yt)
y = yt + stdev*randn(npts)

# RegularizationTools
I = Array{Float64}(LA.I, (npts, npts)) # `A` matrix
Ψ = RT.setupRegularizationProblem(I, 2) 
lmbd = 10.0
ybar = RT.solve(Ψ, y, lmbd) # direct solve in so-called "standard" form
yhat = RT.to_general_form(Ψ, y, ybar)

# direct smoothing by regularization -- given λ and equally spaced x 
function deriv_matrix(d::Int64, N::Int64)
    if d == 0
        return Array{Int64}(LA.I, (N, N))
    end
    return diff(deriv_matrix(d-1, N), dims=1)
end
d = 2
λ = 1e-3 # "scaled" lambda
D = deriv_matrix(d, npts)
δ = LA.tr(D'D)/npts^(d+2)
yhatd = (LA.I + λ/δ*D'D)\y

figure(1)
clf()
plot(x, y, "o", mfc="None")
plot(x, yhat, label="Regularization tools")
plot(x, yhatd, label="Direct method")
legend(loc="best")


examples/scat_example.jl
# Example 2 from manual to build the design matrix 
using RegularizationTools
using Underscores
using MLStyle
using Memoize
using LinearAlgebra
using Gadfly

import Lazy.@>, Lazy.@>>

include("mie_functions.jl")
include("size_functions.jl")

function get_domainfunction(ref)   
    function f(node::Domain)
        s, x, q = node.s, node.x, node.q
        Dp, λ = (@_ map(_[2], s)), q[2]
       
        Q = @match q[1][1:3] begin 
            "βs" => @_ map(Qsca(π*Dp[_]*1e-9/λ, ref), 1:length(Dp))
            "βa" => @_ map(Qabs(π*Dp[_]*1e-9/λ, ref), 1:length(Dp))
            _ => trow("error")
        end
        mapreduce(+, 1:length(Dp)) do i
            π/4.0 *(x[i]*1e6)*Q[i]*(Dp[i]*1e-6)^2.0*1e6
        end
    end
end

Dp, N = lognormal([[100.0, 500.0, 1.4]], d1 = 100.0, d2 = 1000.0, bins = 8)
λs = [300e-9, 500e-9, 900e-9]

s = zip(["Dp$i" for i = 1:length(Dp)],Dp) |> collect
q1 = zip(["βs$i" for i in Int.(round.(λs*1e9, digits=0))], λs) |> collect
q2 = zip(["βa$i" for i in Int.(round.(λs*1e9, digits=0))], λs) |> collect
q = [q1; q2] 

ref = complex(1.8, 0.01)
f = get_domainfunction(ref)
A = designmatrix(s, q, f) 
b = forwardmodel(s, N, q, f) 

r = [0.6640146131466915, 0.2160120227783171, 0.6217884023541196,
	 0.6704021370120161, 0.19721082328252892, 0.3680701880781616]

y = A*N 
b = A*N .+ 0.001*y .* r

xλ = @> setupRegularizationProblem(A,0) solve(b, alg=:gcv_svd) getfield(:x)
plot(x = Dp, y = xλ, Geom.line, layer(x = Dp, y = N))
